ecs-agent/api/ecs/client/ecs_client.go

```diff
@@ -0,0 +1,801 @@
+	registrationAttributes = append(registrationAttributes, client.getAdditionalAttributes()...)
+	registrationAttributes = append(registrationAttributes, client.getOutpostAttribute(outpostARN)...)
+
+	registerRequest.Attributes = registrationAttributes
+	if len(tags) > 0 {
+		registerRequest.Tags = tags
+	}
+	registerRequest.PlatformDevices = platformDevices
+	registerRequest = client.setInstanceIdentity(registerRequest)
+
+	resources, err := client.getResources()
+	if err != nil {
+		return "", "", err
+	}
+
+	registerRequest.TotalResources = resources
+
+	registerRequest.ClientToken = &registrationToken
+	resp, err := client.standardClient.RegisterContainerInstance(&registerRequest)
+	if err != nil {
+		logger.Error("Unable to register as a container instance with ECS", logger.Fields{
+			field.Error: err,
+		})
+		return "", "", err
+	}
+
+	var availabilityzone = ""
+	if resp != nil {
+		for _, attr := range resp.ContainerInstance.Attributes {
+			if aws.StringValue(attr.Name) == azAttrName {
+				availabilityzone = aws.StringValue(attr.Value)
+				break
+			}
+		}
+	}
+
+	logger.Info("Registered container instance with cluster!")
+	err = validateRegisteredAttributes(registerRequest.Attributes, resp.ContainerInstance.Attributes)
+	return aws.StringValue(resp.ContainerInstance.ContainerInstanceArn), availabilityzone, err
+}
+
+func (client *ecsClient) setInstanceIdentity(
+	registerRequest ecsmodel.RegisterContainerInstanceInput) ecsmodel.RegisterContainerInstanceInput {
+	instanceIdentityDoc := ""
+	instanceIdentitySignature := ""
+
+	if client.configAccessor.NoInstanceIdentityDocument() {
+		logger.Info("Fetching Instance ID Document has been disabled")
+		registerRequest.InstanceIdentityDocument = &instanceIdentityDoc
+		registerRequest.InstanceIdentityDocumentSignature = &instanceIdentitySignature
+		return registerRequest
+	}
+
+	iidRetrieved := true
+	backoff := retry.NewExponentialBackoff(setInstanceIdRetryBackoffMin, setInstanceIdRetryBackoffMax,
+		setInstanceIdRetryBackoffJitter, setInstanceIdRetryBackoffMultiple)
+	ctx, cancel := context.WithTimeout(context.Background(), setInstanceIdRetryTimeOut)
+	defer cancel()
+	err := retry.RetryWithBackoffCtx(ctx, backoff, func() error {
+		var attemptErr error
+		logger.Debug("Attempting to get Instance Identity Document")
+		instanceIdentityDoc, attemptErr = client.ec2metadata.GetDynamicData(ec2.InstanceIdentityDocumentResource)
+		if attemptErr != nil {
+			logger.Debug("Unable to get instance identity document, retrying", logger.Fields{
+				field.Error: attemptErr,
+			})
+			// Force credentials to expire in case they are stale but not expired.
+			client.credentialsProvider.Expire()
+			client.credentialsProvider = instancecreds.GetCredentials(client.configAccessor.External())
+			return apierrors.NewRetriableError(apierrors.NewRetriable(true), attemptErr)
+		}
+		logger.Debug("Successfully retrieved Instance Identity Document")
+		return nil
+	})
+	if err != nil {
```

Comment:
Under what scenario, do we return an error? AFAIU, the retry will backoff and continue trying. For containerd agent, I am not sure how it plays out. The agent tries once, returns and sets iidRetrieved false.

Reply:
> Under what scenario, do we return an error?

The function returns an error on L251 if unable to get the instance identity document.

> AFAIU, the retry will backoff and continue trying. For containerd agent, I am not sure how it plays out. The agent tries once, returns and sets iidRetrieved false.

Retry is done with exponential backoff for a maximum of 5 seconds. See L59-L64 and L268-272. Prior to https://github.com/aws/amazon-ecs-agent/pull/3795, Agent only tried once with no retry - so in the event of transient/intermittent upstream failures, this would be a single point of failure in Agent registering with ECS Control Plane.